### PR TITLE
Query cache: Improve handling of system table names in literals

### LINKED ISSUE #69010: Cached queries fail when a string parameter looks like a system table
**Company or project name**

Dash0

**Describe what's wrong**

Whenever a query is being cached (using `SETTINGS use_query_cache = true`) and some string 'looks' like a system table, the query fails with the message:

> Code: 719. DB::Exception: Received from localhost:9000. DB::Exception: The query result was not cached because the query contains a system table. Use setting `query_cache_system_table_handling = 'save'` or `= 'ignore'` to cache the query result regardless or to omit caching. (QUERY_CACHE_USED_WITH_SYSTEM_TABLE)

However no system table is being used. Try e.g. with:

```sql
CREATE TABLE users (uid Int16, name String, age Int16) ENGINE=Memory;

SELECT * FROM users WHERE name = 'system.metrics' SETTINGS use_query_cache = true;
```

Fiddle: https://fiddle.clickhouse.com/efd71f42-006d-4dff-9a2b-3445d1cf9c1f

**Does it reproduce on the most recent release?**

Yes, both `24.8.2` (Fiddle) and `24.6.3` (local version) affected

**Enable crash reporting**

**How to reproduce**

See the above Fiddle

**Expected behavior**

Caching works normally since its a 'regular' SQL statement that gets cached fine with another string value

**Error message and/or stacktrace**

> Code: 719. DB::Exception: Received from localhost:9000. DB::Exception: The query result was not cached because the query contains a system table. Use setting `query_cache_system_table_handling = 'save'` or `= 'ignore'` to cache the query result regardless or to omit caching. (QUERY_CACHE_USED_WITH_SYSTEM_TABLE)

**Additional context**

### PR DESCRIPTION
**Pull request by rschu1ze**
Resolves #69010

### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
SELECT queries run with setting `use_query_cache = 1` are no longer rejected if the name of a system table appears as a literal, e.g. `SELECT * FROM users WHERE name = 'system.metrics' SETTINGS use_query_cache = true;` now works.

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
*This is an automated comment for commit fae5b1170910d8e2b6cd0bf7e12b9a72cbb9bb67 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/71254/fae5b1170910d8e2b6cd0bf7e12b9a72cbb9bb67/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/2c4d93f5523da36e4e7fb491a4e338a7c59e4d28/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: src/Interpreters/Cache/QueryCache.cpp [hunk 1]
```diff
@@ -89,11 +89,40 @@ struct HasSystemTablesMatcher
...
+            {
+                const ASTs & function_children = function->children;
+                if (!function_children.empty())
+                {
+                    if (const auto * expression_list = function_children[0]->as<ASTExpressionList>())
+                    {
+                        const ASTs & expression_list_children = expression_list->children;
+                        if (!expression_list_children.empty())
+                        {
+                            if (const auto * literal = expression_list_children[1]->as<ASTLiteral>())
```
Comment (Avogar) [inline src/Interpreters/Cache/QueryCache.cpp:117]: Let's check that we have at least 2 children. Query `SELECT * FROM clusterAllReplicas(cluster_name)` is valid

Comment (rschu1ze) [reply src/Interpreters/Cache/QueryCache.cpp:117]: Valid point, fixed.

**Review by Avogar — APPROVED** [review]

**Comment by clickgapai** [comment]
Hi @rschu1ze @Avogar — while reviewing this PR I found the following:

- **Bug (P2):** `HasSystemTablesMatcher` misses system tables in `remote`, `cluster`, `remoteSecure` with quoted names — https://github.com/ClickHouse/ClickHouse/issues/104675

Happy to discuss — close anything that's wrong or already addressed.

**Pull request merged**